File: link-list/sum-list/sum-list.js
// Write a function, sumList, that takes in the head of a linked list containing numbers as an argument. The function should return the total sum of all values in the linked list.

class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

const sumList = head => {
    let sum = 0
    while(head !== null) {
        sum += head.val;
        head = head.next;
    }
    return sum
};


const x = new Node(38);
const y = new Node(4);

x.next = y;

// 38 -> 4

console.log(sumList(x)); // 42

File: 2.link-list/3.sum-list/sum-list.go
// Write a function, sumList, that takes in the head of a linked list containing numbers as an argument. The function should return the total sum of all values in the linked list.

package main

import "fmt"

type Node struct {
	value int
	next  *Node
}

func NewNode(value int) *Node {
	return &Node{value: value}
}

// iterative
// func sumList(head *Node) int {
// 	sum := 0
// 	for head != nil {
// 		sum += head.value
// 		head = head.next
// 	}
// 	return sum
// }

// recursive
func sumList(head *Node) int {
	if head == nil {
		return 0
	}
	return head.value + sumList(head.next)
}

func main() {
	x := NewNode(38)
	y := NewNode(4)
	x.next = y

	fmt.Println(sumList(x))
}
